File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/utils/HttpConnections.java
package ru.kmorozov.gbd.utils;

import com.google.api.client.http.GenericUrl;
import com.google.api.client.http.HttpHeaders;
import com.google.api.client.http.HttpResponse;
import com.google.api.client.http.javanet.NetHttpTransport;
import ru.kmorozov.gbd.core.logic.Proxy.HttpHostExt;
import ru.kmorozov.gbd.core.logic.Proxy.UrlType;
import ru.kmorozov.gbd.core.logic.context.BookContext;
import ru.kmorozov.gbd.core.logic.context.ExecutionContext;
import ru.kmorozov.db.core.logic.model.book.google.GoogleBookData;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Created by km on 22.11.2015.
 */
public final class HttpConnections {

    public static final String USER_AGENT = "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 YaBrowser/17.6.1.871 Yowser/2.5 Safari/537.36";
    private static final HttpHeaders headers = new HttpHeaders().setUserAgent(HttpConnections.USER_AGENT);
    private static final HttpConnections INSTANCE = new HttpConnections();
    private static final Map<UrlType, GenericUrl> baseUrls = new HashMap<>(1);

    private final Map<HttpHostExt, HttpHeaders> headersMap = new ConcurrentHashMap<>();

    private HttpConnections() {
    }

    private static Proxy hostToProxy(InetSocketAddress host) {
        return null == host ? null : new Proxy(Proxy.Type.HTTP, host);
    }

    public static HttpResponse getResponse(InetSocketAddress host, final UrlType urlType) {
        return HttpConnections._getResponse(HttpConnections.hostToProxy(host), urlType);
    }

    public static HttpHeaders getHeaders(HttpHostExt proxy) {
        return HttpConnections.INSTANCE._getHeaders(proxy);
    }

    public static String getCookieString(InetSocketAddress proxy, final UrlType urlType) {
        try {
            HttpResponse resp = HttpConnections.getResponse(proxy, urlType);

            if (null == resp) return "";

            return ((Collection<String>) resp.getHeaders().get("set-cookie")).stream()
                    .map(s -> s.split(";")[0]).collect(Collectors.joining(";"));
        } catch (Exception e) {
            return null;
        }
    }

    private static void _setDefaultCookies(Map<String, String> cookiesMap) {
        StringBuilder cookieBuilder = new StringBuilder();

        for (Map.Entry<String, String> cookieEntry : cookiesMap.entrySet()) {
            cookieBuilder.append(cookieEntry.getKey()).append('=').append(cookieEntry.getValue()).append("; ");
        }

        HttpHostExt.NO_PROXY.setCookie(cookieBuilder.toString());
    }

    private HttpHeaders _getHeaders(HttpHostExt proxy) {
        return this.headersMap.computeIfAbsent(proxy, httpHostExt -> {
            HttpHeaders _headers = new HttpHeaders();
            _headers.setUserAgent(HttpConnections.USER_AGENT);
            _headers.setCookie(httpHostExt.getCookie());

            return _headers;
        });
    }

    private static GenericUrl getBaseUrl(final UrlType urlType) {
        GenericUrl url = HttpConnections.baseUrls.get(urlType);
        if (url == null) {
            switch (urlType) {
                case GOOGLE_BOOKS:
                    Optional<BookContext> anyContext = ExecutionContext.INSTANCE.getContexts(false).stream().filter(bookContext -> bookContext.getBookInfo().getBookData() instanceof
                            GoogleBookData).findAny();
                    if (!anyContext.isPresent()) return new GenericUrl("http://www.ya.ru");
                    else {
                        url = new GenericUrl(((GoogleBookData) anyContext.get().getBookInfo().getBookData()).getBaseUrl());
                    }
                    break;
                case JSTOR:
                    url = new GenericUrl("https://www.jstor.org");
                    break;
                default:
                    url = new GenericUrl("http://www.ya.ru");
            }
            HttpConnections.baseUrls.put(urlType, url);
        }
        return url;
    }

    private static HttpResponse _getResponse(Proxy proxy, final UrlType urlType) {
        final GenericUrl baseUrl = HttpConnections.getBaseUrl(urlType);

        try {
            return new NetHttpTransport.Builder().setProxy(proxy).build().createRequestFactory().buildGetRequest(baseUrl).setHeaders(HttpConnections.headers).setConnectTimeout(10000).execute();
        } catch (IOException e) {
            return null;
        }
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/utils/HttpConnections.kt
package ru.kmorozov.gbd.utils

import com.google.api.client.http.HttpHeaders
import ru.kmorozov.gbd.core.config.GBDOptions
import ru.kmorozov.gbd.core.logic.proxy.HttpHostExt
import ru.kmorozov.gbd.core.logic.proxy.TorProxy
import ru.kmorozov.gbd.core.logic.proxy.UrlType
import ru.kmorozov.gbd.logger.Logger
import java.io.IOException
import java.net.*
import java.net.Proxy.NO_PROXY
import java.net.Proxy.Type
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.stream.Collectors

/**
 * Created by km on 22.11.2015.
 */
class HttpConnections private constructor() {

    companion object {

        const val USER_AGENT =
            "Mozilla/5.0 (Linux; Android 7.0; SM-G892A Build/NRD90M; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/60.0.3112.107 Mobile Safari/537.36"
        private val baseUrls = ConcurrentHashMap<UrlType, URI>(1)

        private val logger = Logger.getLogger(GBDOptions.debugEnabled, HttpConnections::class.java)

        fun getResponse(host: InetSocketAddress, urlType: UrlType): Optional<HttpResponse<*>> {
            val baseUrl = getBaseUrl(urlType)
            val proxy: Proxy
            if ("localhost" == host.hostName && 1 == host.port)
                proxy = NO_PROXY
            else if (TorProxy.TOR_HOST == host.hostName && 9150 == host.port)
                proxy = Proxy(Type.SOCKS, InetSocketAddress(TorProxy.TOR_HOST, TorProxy.TOR_HTTP_PORT))
            else
                proxy = Proxy(Type.HTTP, host)

            try {
                val client = if (proxy == NO_PROXY) HttpClient.newBuilder().build() else HttpClient.newBuilder()
                    .proxy(ProxySelector.of(host)).build()
                val reqBuilder = HttpRequest.newBuilder()
                    .uri(baseUrl).GET().timeout(Duration.ofMillis(10000)).setHeader("User-Agent", USER_AGENT)

                return Optional.of(client.send(reqBuilder.build(), HttpResponse.BodyHandlers.ofInputStream()))
            } catch (e: IOException) {
                if (GBDOptions.debugEnabled)
                    logger.info("Proxy $proxy error: " + e.message)

                return Optional.empty()
            } catch (ste: SocketTimeoutException) {
                if (GBDOptions.debugEnabled) {
                    logger.info("Proxy $proxy timeout")
                }

                return Optional.empty()
            } catch (ce: ConnectException) {
                if (GBDOptions.debugEnabled) {
                    logger.info("Cannot connect $proxy")
                }

                return Optional.empty()
            }
        }

        fun getHeaders(proxy: HttpHostExt): HttpHeaders {
            val headers = HttpHeaders()
            headers.userAgent = USER_AGENT
            headers.cookie = proxy.cookie
            return headers
        }

        fun getCookieString(proxy: InetSocketAddress, urlType: UrlType): String {
            val resp = getResponse(proxy, urlType)
            if (resp.isEmpty)
                return ""
            else {
                if (GBDOptions.debugEnabled)
                    logger.info("Got cookies for $proxy : ${resp.get().headers()}")

                return resp.get().headers().allValues("set-cookie").stream()
                    .map { s -> s.split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0] }
                    .collect(Collectors.joining(";"))
            }
        }

        private fun getBaseUrl(urlType: UrlType): URI {
            return baseUrls.computeIfAbsent(urlType) {
                when (urlType) {
                    UrlType.GOOGLE_BOOKS -> URI.create("https://www.google.com/")
                    UrlType.GOOGLE_BOOK_INFO -> URI.create("https://www.google.com/")
                    UrlType.JSTOR -> URI.create("https://www.jstor.org")
                    else -> URI.create("http://www.ya.ru")
                }
            }
        }
    }
}
